### ISSUE TITLE
Connect domain caches

### ISSUE THREAD

**Issue by philipmarnef** [getkirby/kirby#2248]
Ran into an issue where a site was accessible on both www.example.com and example.com and only one of the caches was getting flushed depending on the domain where the panel was accessed.

This can be avoided by a redirect, but I can think of a few situations where you would want to use a few subdomains on one Kirby installation and flush the cache for all domains automatically when content changes.

AFAIK there is currently no easy way to achieve this.

I'm thinking of an 'invalidate' property under the config cache option where you can list which domain caches to flush if something is updated in the panel.

**Comment by lukasbestle** [comment]
I actually think that this is a bug. Kirby should probably clear all page caches on every content update, not just the one of the current domain. Or would there be use-cases where you would want the caches of other domains to stay intact?

**Comment by philipmarnef** [comment]
I can think of many examples where you _don't_ want the cache to stay intact, but I haven't thought about the opposite scenario. I suppose it depends on the question whether Kirby should support a multi-site installation (which is afaik not really the intention)? But even then the default behaviour should probably be all caches cleared.

**Comment by philipmarnef** [comment]
I just learned Kirby allows a multi-site setup, but that implies a different cache folder for each site. So definitely a bug IMO.

**Comment by neildaniels** [comment]
This is almost certainly caused by the fix for https://github.com/getkirby/kirby/issues/1496 a while ago. In that case, content was getting commingled between a dev and production server.

I _think_ (but haven't tested) the "correct" way to specially opt into cache sharing is to set a `prefix` for in your cache config.

By default, the `prefix` is the [index URL](https://getkirby.com/docs/reference/system/urls/index) (which in your case, differs for `example.com` and `www.example.com`).

```php
return [
  'cache' => [
      'pages' => [
          'active' => true,
          'prefix' => 'example.com',
      ]
  ]
];
```

But do note, this means page-level caches will be mixed from both domains. For example, if the first users to a `/foo` page visits as `www.example.com/foo`, and that page is cached, then another user that visits `example.com/foo` will receive a page that links to `www.example.com` pages. This might not functionally make a difference, but considering how it can be unpredictable, it's probably not ideal.

(If you're providing canonical URLs, and they're varying depending on which domain you're on, you're probably also confusing indexers.)

Alternatively, you might consider setting a "preferred" [index URL](https://getkirby.com/docs/reference/system/urls/index) so that all links on the page are generated as `example.com` **or** `www.example.com`. This should also automatically enable a shared cache (but again, I'm saying that untested).

With regards to multi-lang site:
I don't use a multi-lang site, and I have no idea what happens in relation to this.

**Comment by philipmarnef** [comment]
Ok, so the easiest solutions seem to be either
1. the docs should explain that in case of domain aliases a canonical domain should be set in `index.php` and as cache prefix in `config.php` (the latter is currently not documented)
OR
2. a patch is made so that `->flush()` clears ALL caches on the same installation at once

**Comment by neildaniels** [comment]
@philipmarnef I think that's the right direction.

I'd add the following:
1. You should only need to set the canonical domain **or** set the cache prefix. Setting the canonical domain should remove the need to set a cache prefix.
2. I think `->flush()` should **not** _automatically_ clear all caches on an installation, I do think it would be incredibly helpful to be able to optionally configure "flush dependencies" that should also get cleared when a cache gets cleared.
   - For example: We a custom cache bucket that should get cleared out whenever a change to pages is made. Right now, we have to listen for page actions and clear it ourself. It would be much easier to say "when the regular `pages` cache gets flushed, also flush this one`.

**Comment by lukasbestle** [comment]
@neildaniels We'd only clear all *pages* caches of all domains as those are definitely all invalidated if content changes. Custom caches would be unaffected. The cache dependency idea is another separate feature.

**Comment by neildaniels** [comment]
@lukasbestle Gotcha. I'm mostly concerned that if I make content changes to `dev` I wouldn't want that to invalidate my production caches, because those point to completely different `content` and `storage` folders (as specified in my `index.php`).

I think automatically clearing other `pages` caches is fine, as long as their `content` folders are actually the same.

**Comment by lukasbestle** [comment]
Well, but if you have completely different content folders, you can also have completely different cache folders, right?

**Comment by neildaniels** [comment]
Yes, but even if they shared the same `content` and `storage` folders, they'd still technically be in separate folders.

```
- storage
  - cache
    - example.com
      - pages
    - www.example.com
      - pages
    - 10.0.0.1
      - pages
```

I'm honestly not sure "how" the `example.com` should/would know that `www.example.com` and `10.0.0.1` are the "same" site (for purposes of cache clearing).

I also think a _lot_ of assumptions are being made here that make this impractical to completely automate.
 - What if one domain uses APCu caching, but the other uses file cache.
 - The other domains might have separate cache prefixes, which would _only_ be loaded if that site is actively being loaded.

Simply put: I don't see it being feasible to try and "know" the cache setup of other domains, when those configurations haven't been loaded.

**Comment by lukasbestle** [comment]
If the `cache` folder is the same, we assume that it's the cache for the same site. The separate domains inside each cache folder all belong to the same site. The reason we store the caches per domain is to enable multi-domain multilang setups.

> What if one domain uses APCu caching, but the other uses file cache.

That's indeed a huge problem. We would need to load all domain-specific config files, which is impractical.

To be honest, I think that this is something we can't fix. Manually setting a common cache prefix is probably the way to go to connect the caches. Even better would be to redirect one domain to the other – this also avoids duplicate content.

**Comment by iskrisis** [comment]
I run to same thing with www. non-www domains. And while it was confusing for a bit... i've discovered that i have set up redirects in a wrong way because of it. If it was flushing all caches i would have never noticed. I don't think anybody really wants www and non-www running at the same time. It's always a mistake.

So to me the way it works currently seems fine. Maybe some mention of this in docs. Or even just linking to this issue in note in docs would be helpful.

**Comment by distantnative** [comment]
@lukasbestle Reading the discussion again, I would agree that there isn't a good way to fix this and thus should remain in its current version. If you still think this as well, please close as `unsolvable`.

**Comment by bastianallgeier** [comment]
We already have a todo for the prefix documentation to fix this. I agree that we should close this.
